lib/datadog/appsec/event.rb

```diff
@@ -43,6 +43,9 @@ def self.record(*events)
       end
 
       # rubocop:disable Metrics/AbcSize
+      # rubocop:disable Metrics/MethodLength
+      # rubocop:disable Metrics/PerceivedComplexity
+      # rubocop:disable Metrics/CyclomaticComplexity
       def self.record_via_span(*events)
         events.group_by { |e| e[:trace] }.each do |trace, event_group|
           unless trace
```

Comment:
@lloeki The sinatra test failure `runner           | 17:39:00.895 ERROR    Trace id 4359199009121235128 sampling priority is 2.0, should be (0, -1)` happens when `trace.keep!` is invoked on line 56 of this file👇.

The span was not sampled (due to the sampling rate being 50%, and trace 4359199009121235128 being "randomly" selected to be dropped), but calling `trace.keep!` overrides that.

This is the same discussion we are having in the other Appsec PR.

The question here is: do you want Appsec to keep an otherwise discarded span?
In this specific case, the sampling rate went from -1 (USER_REJECT) to 2 (USER_KEEP) in the Sinatra test case, so it was previously set to "explicitly rejected by user".

Reply:
Unfortunately the issue is that this `trace.keep!` is only called if an appsec event is to be recorded, yet the failing test does a very simple `30.times { |i| GET /sample_rate_route/#{i} }` which does not trigger any event (on purpose, this is in fact not an appsec system test case but a tracing one).

What is confounding is that just by adding a couple of `Datadog.logger.info` calls to log conditions leading to a call to `record` in `DispatchPatch#dispatch!` relevant values are `nil` as expected and the test passes again, reliably, and removing these makes it fail again, reliably.

It feels kind of like a timing issue, so I'm wondering if this has a tie with AppSec rate limiting (defaults to 100 traces/s/thread, and since there are many requests in addition to those 30 ones the limit might get hit) but I don't see how that would make the value `2.0` either, unless the tests get confused somehow.

In addition `inspect` (and thus span debug logs when diagnostics are enabled) does not seem to show all relevant values anymore (e.g it never shows `_sampling_priority_v1`) which makes it annoying to debug. I suppose this is due to these values now being stored on the trace and it only logs spans prior to them being wire-formatted but not trace-level information.

Coincidentally I just realised I should probably amend `#record` with an early `return if events.empty?` guard clause, otherwise with a lacking condition or unexpected written by the caller, rate limiting would still happen even when there is no event to record:

```ruby
      def self.record(*events)
        Datadog::AppSec::RateLimiter.limit(:traces) do
          record_via_span(*events)
        end
      end
```

Reply:
Alright, got to the bottom of it:

```
runner           | 23:18:40.997 ERROR    Trace id 3628486659673762649 sampling priority is 2.0, should be (0, -1)
```

indeed comes from:

```
runner           | E           interfaces.library.assert_sampling_decision_respected(context.sampling_rate)
```

But the request that triggers it is not one of the 30 requests done in that test. Instead, following `3628486659673762649` in `logs/docker/weblog/stdout.log`:

```
D, [2022-05-05T23:17:57.430648 #11] DEBUG -- ddtrace: [ddtrace] (/binaries/dd-trace-rb/lib/datadog/appsec/reactive/operation.rb:15:in `initialize') operation: rack.request initialize
D, [2022-05-05T23:17:57.430919 #11] DEBUG -- ddtrace: [ddtrace] (/binaries/dd-trace-rb/lib/datadog/appsec/contrib/rack/reactive/request.rb:38:in `block in subscribe') reacted to ["request.headers", "request.uri.raw", "request.query", "request.cookies"]: [{"version"=>"HTTP/1.1", "host"=>"weblog:7777", "accept-encoding"=>"identity", "user-agent"=>"system_tests rid/HGKMNGMXCGXIZILXDOJNQHWCYBUYDDYJAHGB"}, "http://weblog:7777/waf", [], {}]
D, [2022-05-05T23:17:57.438227 #11] DEBUG -- ddtrace: [ddtrace] (/binaries/dd-trace-rb/lib/datadog/appsec/contrib/rack/reactive/request.rb:71:in `block in subscribe') WAF OK: #<struct Datadog::AppSec::WAF::Result action=:good, data=nil, total_runtime=119302, timeout=false>
D, [2022-05-05T23:17:57.438430 #11] DEBUG -- ddtrace: [ddtrace] (/binaries/dd-trace-rb/lib/datadog/appsec/reactive/operation.rb:39:in `finalize') operation: rack.request finalize
D, [2022-05-05T23:17:57.439172 #11] DEBUG -- ddtrace: [ddtrace] (/binaries/dd-trace-rb/lib/datadog/appsec/reactive/operation.rb:15:in `initialize') operation: rack.request.body initialize
D, [2022-05-05T23:17:57.440486 #11] DEBUG -- ddtrace: [ddtrace] (/binaries/dd-trace-rb/lib/datadog/appsec/contrib/rack/reactive/request_body.rb:27:in `block in subscribe') reacted to ["request.body"]: [["<vmlframe src=\"xss\">"]]
....
[
 Name: sinatra.request
Span ID: 2000327487103858201
Parent ID: 2988872142661438865
Trace ID: 3628486659673762649
Type: web
Service: weblog
Resource: POST /waf
```

And then following `HGKMNGMXCGXIZILXDOJNQHWCYBUYDDYJAHGB` in `logs/pytest.log`:

```
23:17:57.426 DEBUG    Sending request HGKMNGMXCGXIZILXDOJNQHWCYBUYDDYJAHGB: POST http://weblog:7777/waf
23:17:57.531 DEBUG    Request HGKMNGMXCGXIZILXDOJNQHWCYBUYDDYJAHGB: 500
23:17:57.539 DEBUG    Interface: None -> _WafAttack: [94mAppSec detects attacks in JSON body arrays[0m is called from <class 'tests.appsec.waf.test_addresses.Test_BodyJson'>, which is xfail
23:17:57.540 DEBUG    _WafAttack('AppSec detects attacks in JSON body arrays', HGKMNGMXCGXIZILXDOJNQHWCYBUYDDYJAHGB) added in library interface[70]
```

So, this returns a 500, confirmed by:

```
Name: rack.request
Span ID: 2988872142661438865
Parent ID: 0
Trace ID: 3628486659673762649
Type: web
Service: weblog
Resource: POST /waf
Error: 1
Start: 1651792677430393856
End: 1651792677529313024
Duration: 0.09891891200095415
Tags: [
   key1 => val1,
   key2  =>  val2,
   env => system-tests,
   component => rack,
   operation => request,
   http.method => POST,
   http.url => /waf,
   http.base_url => http://weblog:7777/,
   http.status_code => 500,
```

And this is the error:

```
   error.type => TypeError,
   error.msg => no implicit conversion of Array into Hash,
   error.stack => /usr/local/bundle/gems/rack-2.2.3/lib/rack/request.rb:469:in `merge': no implicit conversion of Array into Hash (TypeError)
	from /usr/local/bundle/gems/rack-2.2.3/lib/rack/request.rb:469:in `params'
	from /usr/local/bundle/gems/rack-2.2.3/lib/rack/request.rb:32:in `params'
	from /usr/local/bundle/gems/sinatra-2.1.0/lib/sinatra/base.rb:78:in `params'
	from /usr/local/bundle/gems/sinatra-2.1.0/lib/sinatra/base.rb:1118:in `dispatch!'
	from /binaries/dd-trace-rb/lib/datadog/appsec/contrib/sinatra/patcher.rb:57:in `block in dispatch!'
	from /binaries/dd-trace-rb/lib/datadog/appsec/instrumentation/gateway.rb:21:in `block in push'
	from /binaries/dd-trace-rb/lib/datadog/appsec/contrib/rack/gateway/watcher.rb:131:in `block in watch'
	from /binaries/dd-trace-rb/lib/datadog/appsec/instrumentation/gateway.rb:27:in `block (2 levels) in push'
	from /binaries/dd-trace-rb/lib/datadog/appsec/instrumentation/gateway.rb:31:in `push'
	from /binaries/dd-trace-rb/lib/datadog/appsec/contrib/sinatra/patcher.rb:56:in `dispatch!'
	from /usr/local/bundle/gems/sinatra-2.1.0/lib/sinatra/base.rb:939:in `block in call!'
	from /usr/local/bundle/gems/sinatra-2.1.0/lib/sinatra/base.rb:1101:in `block in invoke'
	from /usr/local/bundle/gems/sinatra-2.1.0/lib/sinatra/base.rb:1101:in `catch'
	from /usr/local/bundle/gems/sinatra-2.1.0/lib/sinatra/base.rb:1101:in `invoke'
	from /usr/local/bundle/gems/sinatra-2.1.0/lib/sinatra/base.rb:939:in `call!'
	from /usr/local/bundle/gems/sinatra-2.1.0/lib/sinatra/base.rb:929:in `call'
	from /usr/local/bundle/gems/rack-contrib-2.3.0/lib/rack/contrib/json_body_parser.rb:69:in `call'
```

Turns out this test tries to post `[]`, which `Rack::Contrib`'s `JSONBodyParser` happily converts, but then Sinatra expects a `Hash`. It does not fail with `Rack` because the small Rack app doesn't attempt to do anything with it. 

This behaviour is completely outside of the scope of `ddtrace`. I will disable the array test for Sinatra for now.

@marcotc how ironic that it is relevant to your comment here: https://github.com/DataDog/dd-trace-rb/pull/1969#discussion_r866374369

Reply:
> I will disable the array test for Sinatra for now.

Done: https://github.com/DataDog/system-tests/pull/342